services/src/main/java/org/keycloak/protocol/ciba/decoupledauthn/DelegateDecoupledAuthenticationProvider.java

```diff
@@ -0,0 +1,281 @@
+        UserModel user = null;
+        if (authRequestedUserHint.equals(CIBAConstants.LOGIN_HINT)) {
+            user = resolver.getUserFromLoginHint(request.getLoginHint());
+        } else if (authRequestedUserHint.equals(CIBAConstants.ID_TOKEN_HINT)) {
+            user = resolver.getUserFromLoginHint(request.getIdTokenHint());
+        } else if (authRequestedUserHint.equals(CIBAConstants.LOGIN_HINT_TOKEN)) {
+            user = resolver.getUserFromLoginHint(request.getLoginHintToken());
+        } else {
+            throw new RuntimeException("CIBA invalid Authentication Requested User Hint.");
+        }
+        String infoUsedByAuthentication = resolver.getInfoUsedByAuthentication(user);
+
+        StringBuilder scopeBuilder = new StringBuilder();
+        Map<String, ClientScopeModel> defaultScopesMap = client.getClientScopes(true, true);
+        defaultScopesMap.forEach((key, value)->{if (value.isDisplayOnConsentScreen()) scopeBuilder.append(value.getName()).append(" ");});
+        String defaultClientScope = scopeBuilder.toString();
+
+        CIBAAuthReqId decoupledAuthIdJwt = new CIBAAuthReqId();
+        decoupledAuthIdJwt.id(KeycloakModelUtils.generateId());
+        decoupledAuthIdJwt.setScope(request.getScope());
+        decoupledAuthIdJwt.setSessionState(userSessionIdWillBeCreated);
+        decoupledAuthIdJwt.setAuthResultId(authResultId);
+        decoupledAuthIdJwt.issuedNow();
+        decoupledAuthIdJwt.issuer(Urls.realmIssuer(session.getContext().getUri().getBaseUri(), realm.getName()));
+        decoupledAuthIdJwt.audience(decoupledAuthIdJwt.getIssuer());
+        decoupledAuthIdJwt.subject(user.getId());
+        decoupledAuthIdJwt.exp(Long.valueOf(Time.currentTime() + expiresIn));
+        //decoupledAuthIdJwt.issuedFor(Decoupled Auth Server's client_id); TODO
+        decoupledAuthIdJwt.issuedFor(client.getClientId()); // TODO : set CD's client_id intentionally, not Decoupled Auth Server. It is not good idea so that client_id field should be added.
+        String decoupledAuthId = CIBAAuthReqIdParser.persistAuthReqId(session, decoupledAuthIdJwt);
+
+        logger.info("  decoupledAuthnRequestUri = " + decoupledAuthenticationRequestUri);
+        try {
+            int status = SimpleHttp.doPost(decoupledAuthenticationRequestUri, session)
+                .param(DECOUPLED_AUTHN_ID, decoupledAuthId)
+                .param(DECOUPLED_AUTHN_USER_INFO, infoUsedByAuthentication)
+                .param(DECOUPLED_AUTHN_IS_CONSENT_REQUIRED, Boolean.toString(client.isConsentRequired()))
+                .param(CIBAConstants.SCOPE, request.getScope())
+                .param(DECOUPLED_DEFAULT_CLIENT_SCOPE, defaultClientScope)
+                .param(CIBAConstants.BINDING_MESSAGE, request.getBindingMessage())
+                .asStatus();
+            logger.info("  Decoupled Authn Request URI Access = " + status);
+            if (status != 200) {
```

Comment:
In case you have a status other than `200`, such as `404`, `500`, or any other that indicates that the chosen channel did not receive the authentication request, shouldn't we fail?

Also, the contract is not clear to me in regards to the expected response from the channel. Wouldn't be better to enforce a specific status code? Probably `201` (created) ?

Reply:
@pedroigor As you said, we would want to fail the backchannel authentication request when we received some error response from the decoupled authentication server, namely received status code other than 200. However, at this time, we has already sent the backchannel authentication response to the client (consumption device) so that we could not tell the client this error response.

As for the status code for successful response from the decoupled authentication server, I agree with you. I would like to use 201 instead of 200.